src/Language/Haskell/Refact/Refactoring/GenApplicative.hs

```diff
@@ -0,0 +1,383 @@
+      fetchAnnsFinal >>= putRefactParsed newParsed
+  where worker :: GHC.RdrName -> ParsedBind -> RefactGhc (GHC.HsBind GHC.RdrName)
+#if __GLASGOW_HASKELL__ <= 710
+        worker rNm fBind@(GHC.FunBind (GHC.L _ fNm) _ mg _ _ _)
+#else
+        worker rNm fBind@(GHC.FunBind (GHC.L _ fNm) mg _ _ _)
+#endif
+          | fNm == rNm = do
+              newMg <- replaceMG mg
+              return $ fBind{GHC.fun_matches = newMg}
+          | otherwise = return fBind
+        worker _ bind = return bind
+        replaceMG :: ParsedMatchGroup -> RefactGhc ParsedMatchGroup
+        replaceMG mg = do
+#if __GLASGOW_HASKELL__ <= 710
+          let [(GHC.L l match)] = GHC.mg_alts mg
+#else
+          let (GHC.L _ [(GHC.L l match)]) = GHC.mg_alts mg
+#endif
+              oldGrhss = GHC.m_grhss match
+              newGrhss = mkGrhss oldGrhss newRhs
+              newLMatch = (GHC.L l (match{GHC.m_grhss = newGrhss}))
+#if __GLASGOW_HASKELL__ <= 710              
+          return mg{GHC.mg_alts = [newLMatch]}
+#else
+          lMatchLst <- locate [newLMatch]
+          return mg{GHC.mg_alts = lMatchLst}
+#endif
+        mkGrhss old newExpr = let [(GHC.L l (GHC.GRHS lst _))] = GHC.grhssGRHSs old in
+          old{GHC.grhssGRHSs = [(GHC.L l (GHC.GRHS lst newExpr))]}              
+                                   
+          
+
+processReturnStatement :: ParsedExpr -> [GHC.RdrName] -> RefactGhc (Maybe ParsedLExpr)
+processReturnStatement retExpr boundVars
+  | isJustBoundVar retExpr boundVars = return Nothing
+  | otherwise =
+      case retExpr of
+        (GHC.ExplicitTuple lst _) -> do
+          dFlags <- GHC.getSessionDynFlags
+          let commas = repeat ','
+              constr = "(" ++ (take ((length lst)-1) commas) ++ ")"
+              parseRes = parseExpr dFlags "hare" constr
+          case parseRes of
+            (Left (_, errMsg)) -> do
+              logm "processReturnStatement: error parsing tuple constructor"
+              return Nothing
+            (Right (anns, expr)) -> do
+              mergeAnns anns
+              return (Just expr)
+        _ -> do
+          lRet <- locate retExpr
+          stripBoundVars lRet boundVars
+      where stripBoundVars :: ParsedLExpr -> [GHC.RdrName] -> RefactGhc (Maybe ParsedLExpr)
+#if __GLASGOW_HASKELL__ <= 710
+            stripBoundVars le@(GHC.L l (GHC.HsVar nm)) names =
+#else
+            stripBoundVars le@(GHC.L l (GHC.HsVar (GHC.L _ nm))) names =
+#endif
+              if (elem nm names)
+              then return Nothing
+              else return (Just le)
+            stripBoundVars (GHC.L l (GHC.HsApp expr1 expr2)) names = do
+              ne1 <- stripBoundVars expr1 names
+              ne2 <- stripBoundVars expr2 names
+              case ne2 of
+                Nothing -> return ne1
+                (Just e2) -> return (ne1 >>= (\e1 -> Just (GHC.L l (GHC.HsApp e1 e2))))
+
+mergeAnns :: Anns -> RefactGhc ()
```

Comment:
I can find setRefactAnns in MonadFunctions that replaces all of the annotations, there is a function called mergeAnns in Transform in exactprint but that is just a synonym of Map.union. I will rename this function mergeRefactAnns.